File: Dart/14-Backtracking/Problems/3-CombinationSum/CombinationSumPro.dart
// Advanced Dart solution for "Combination Sum" with real-world optimizations

import 'dart:collection';

void main() {
  List<int> candidates = [2, 3, 5, 7];
  int target = 8;

  print('Backtracking with Pruning: ${combinationSumBacktracking(candidates, target)}');
  print('Memoization with State Encoding: ${combinationSumMemoStateEncoding(candidates, target)}');
  print('DP with Space Optimization: ${combinationSumOptimizedDP(candidates, target)}');
  print('BFS with Level Processing: ${combinationSumBFSLevelProcessing(candidates, target)}');
  print('Parallel Processing (Simulated): ${combinationSumParallelProcessing(candidates, target)}');
  print('Hybrid Approach: ${combinationSumHybrid(candidates, target)}');
}

// Approach 1: Backtracking with Pruning
// This approach uses backtracking to explore all possible combinations.
// It sorts the candidates to enable pruning, which significantly reduces the search space.
List<List<int>> combinationSumBacktracking(List<int> candidates, int target) {
  List<List<int>> result = [];
  candidates.sort(); // Optimization: Sort candidates to prune early
  void backtrack(int start, int target, List<int> path) {
    if (target == 0) {
      result.add(List.from(path)); // Found a valid combination
      return;
    }
    for (int i = start; i < candidates.length; i++) {
      if (candidates[i] > target) break; // Pruning: If current candidate > target, no need to continue
      path.add(candidates[i]);
      backtrack(i, target - candidates[i], path); // Recursive call
      path.removeLast(); // Backtrack: Remove the last added candidate
    }
  }
  backtrack(0, target, []);
  return result;
}

// Approach 2: Memoization with State Encoding
// This approach uses memoization to store intermediate results, avoiding redundant calculations.
// The state is encoded as a string to use as a key in the memoization map.
Map<String, List<List<int>>> memo = {};
List<List<int>> combinationSumMemoStateEncoding(List<int> candidates, int target, [int start = 0]) {
  String key = '$start-$target'; // State encoding: start index and remaining target
  if (memo.containsKey(key)) return memo[key]!; // Return memoized result if available
  if (target == 0) return [[]]; // Base case: target reached
  if (target < 0) return []; // Base case: target exceeded
  List<List<int>> result = [];
  for (int i = start; i < candidates.length; i++) {
    for (var sublist in combinationSumMemoStateEncoding(candidates, target - candidates[i], i)) {
      result.add([candidates[i], ...sublist]); // Add current candidate to sublist
    }
  }
  return memo[key] = result; // Store and return the result
}

// Approach 3: Dynamic Programming with Space Optimization
// This approach uses dynamic programming to build solutions bottom-up.
// It optimizes space by storing only the necessary intermediate results.
List<List<int>> combinationSumOptimizedDP(List<int> candidates, int target) {
  List<List<List<int>>> dp = List.generate(target + 1, (_) => []); // DP table: dp[i] stores combinations for target i
  dp[0] = [[]]; // Base case: target 0 has an empty combination
  for (int num in candidates) {
    for (int i = num; i <= target; i++) {
      for (var combination in dp[i - num]) {
        dp[i].add([...combination, num]); // Add current candidate to existing combinations
      }
    }
  }
  return dp[target]; // Return combinations for the target
}

// Approach 4: BFS with Level Processing
// This approach uses Breadth-First Search (BFS) to explore combinations level by level.
// It uses a queue to manage the search process.
List<List<int>> combinationSumBFSLevelProcessing(List<int> candidates, int target) {
  List<List<int>> result = [];
  Queue<Map<String, dynamic>> queue = Queue();
  queue.add({'sum': 0, 'path': <int>[]}); // Initialize queue with starting state
  while (queue.isNotEmpty) {
    var current = queue.removeFirst();
    int sum = current['sum'] as int;
    List<int> path = List<int>.from(current['path']);
    for (int num in candidates) {
      if (path.isNotEmpty && num < path.last) continue; // Avoid duplicate combinations
      int newSum = sum + num;
      if (newSum == target) {
        result.add([...path, num]); // Found a valid combination
      } else if (newSum < target) {
        queue.add({'sum': newSum, 'path': [...path, num]}); // Enqueue next state
      }
    }
  }
  return result;
}

// Approach 5: Parallel Processing (Simulated)
// This approach simulates parallel processing using a recursive worker function.
// In a real-world scenario, this could be implemented using isolates or threads.
List<List<int>> combinationSumParallelProcessing(List<int> candidates, int target) {
  List<List<int>> result = [];
  void worker(List<int> subset, int remaining, int start) {
    if (remaining == 0) {
      result.add(List.from(subset)); // Found a valid combination
      return;
    }
    for (int i = start; i < candidates.length; i++) {
      if (candidates[i] > remaining) break;
      subset.add(candidates[i]);
      worker(subset, remaining - candidates[i], i);
      subset.removeLast();
    }
  }
  worker([], target, 0);
  return result;
}

// Approach 6: Hybrid Approach (Combining DP and Backtracking)
// This approach combines dynamic programming and backtracking for efficiency.
// It uses DP to store intermediate results and backtracking to explore combinations.
List<List<int>> combinationSumHybrid(List<int> candidates, int target) {
  Map<int, List<List<int>>> dp = {}; // DP table: dp[sum] stores combinations for sum
  void backtrack(int sum, List<int> path) {
    if (sum == target) {
      dp[sum] = dp[sum] ?? [];
      dp[sum]!.add(List.from(path));
      return;
    }
    for (int num in candidates) {
      if (sum + num > target) break;
      path.add(num);
      backtrack(sum + num, path);
      path.removeLast();
    }
  }
  backtrack(0, []);
  return dp[target] ?? [];
}

File: Python/14-Backtracking/Problems/3-CombinationSum/CombinationSumPro.py
"""
Combination Sum Problem:

Given a set of candidate numbers (candidates) and a target number (target),
find all unique combinations in candidates where the numbers sum to target.
The same number may be chosen from candidates an unlimited number of times.

For example:
    candidates = [2, 3, 6, 7], target = 7
    Output: [[7], [2, 2, 3]]

    candidates = [2, 3, 5], target = 8
    Output: [[2, 2, 2, 2], [2, 3, 3], [3, 5]]

    candidates = [2], target = 1
    Output: []
"""

def combinationSum_recursive(candidates, target):
    """
    1.  Recursive Approach (Basic Backtracking)

    This is the most straightforward backtracking solution.  It explores all possible
    combinations by recursively adding candidates to the current combination
    and checking if the sum equals the target.

    Time Complexity: O(N^(T/M)), where N is the number of candidates, T is the target,
                    and M is the minimum value in candidates.  This is a loose upper bound.
    Space Complexity: O(T/M) -  The maximum depth of the recursion.
    """
    result = []

    def backtrack(combination, remaining, start):
        if remaining == 0:
            result.append(list(combination))  # Append a *copy* to avoid modification issues
            return
        elif remaining < 0:
            return

        for i in range(start, len(candidates)):
            combination.append(candidates[i])
            backtrack(combination, remaining - candidates[i], i)  # Allow reuse of the same element
            combination.pop()  # Backtrack: remove the last element to try another path

    backtrack([], target, 0)
    return result


def combinationSum_optimized(candidates, target):
    """
    2. Optimized Recursive Approach with Sorting

    This approach improves the basic recursion by sorting the candidates first.
    This allows us to prune the search space earlier when the current sum exceeds
    the target, avoiding unnecessary recursive calls.

    Time Complexity: O(N*logN + N^(T/M)), where N is the number of candidates, T is the target,
                    and M is the minimum value in candidates. The N*logN comes from sorting.
    Space Complexity: O(T/M) - The maximum depth of the recursion.
    """
    result = []
    candidates.sort()  # Sort candidates to optimize

    def backtrack(combination, remaining, start):
        if remaining == 0:
            result.append(list(combination))
            return
        elif remaining < 0:
            return

        for i in range(start, len(candidates)):
            # Optimization: If current candidate is too large, skip it and the rest (since sorted)
            if candidates[i] > remaining:
                break
            combination.append(candidates[i])
            backtrack(combination, remaining - candidates[i], i)
            combination.pop()

    backtrack([], target, 0)
    return result



def combinationSum_iterative(candidates, target):
    """
    3. Iterative Approach using a Stack

    This approach uses a stack to simulate the recursive calls.  It can be
    more complex to understand but avoids the overhead of recursion.

    Time Complexity:  O(N^(T/M)), where N is the number of candidates, T is the target,
                    and M is the minimum value in candidates.
    Space Complexity: O(T/M),  The maximum size of the stack.
    """
    result = []
    stack = [(0, [], target)]  # (start index, current combination, remaining target)

    while stack:
        start, combination, remaining = stack.pop()
        if remaining == 0:
            result.append(list(combination))
            continue
        elif remaining < 0:
            continue

        for i in range(start, len(candidates)):
            stack.append((i, combination + [candidates[i]], remaining - candidates[i]))
    return result



def combinationSum_dp(candidates, target):
    """
    4. Dynamic Programming (DP) Approach (Less Common, Modified)

    While the classic Combination Sum problem doesn't lend itself perfectly to
    traditional DP due to the allowance of repeated numbers, we can adapt a DP-like
    approach to build up solutions.  This is less efficient than backtracking
    but demonstrates a different problem-solving technique.  This approach is
    modified to fit the problem.

    Time Complexity: O(T * N * (T/M)), where T is the target, N is the number of candidates, and M is the smallest candidate.
    Space Complexity: O(T * (T/M)),  where T is the target and M is the smallest candidate.
    """
    dp = [[] for _ in range(target + 1)]  # dp[i] stores combinations that sum to i
    dp[0] = [[]]  # Base case: sum of 0 is an empty combination

    for num in candidates:
        for i in range(num, target + 1):
            for comb in dp[i - num]:
                dp[i].append(comb + [num])
    return dp[target]



def combinationSum_generator(candidates, target):
    """
    5.  Generator-Based Approach (Elegant Backtracking)

    This approach uses a generator function to produce the combinations lazily.
    It's a more Pythonic way to implement backtracking, often considered more readable.

    Time Complexity: O(N^(T/M)), where N is the number of candidates, T is the target,
                    and M is the minimum value in candidates.
    Space Complexity: O(T/M) -  Due to the recursion depth.  The generator itself has minimal overhead.
    """
    def generate_combinations(combination, remaining, start):
        if remaining == 0:
            yield list(combination)
            return
        elif remaining < 0:
            return

        for i in range(start, len(candidates)):
            yield from generate_combinations(combination + [candidates[i]], remaining - candidates[i], i)

    return list(generate_combinations([], target, 0))  # Consume the generator and return a list


def combinationSum_memoized(candidates, target):
    """
    6. Memoized Recursive Approach (Top-Down DP)

    This approach uses memoization to store the results of subproblems, effectively
    turning the recursive solution into a top-down dynamic programming approach.
    This can improve performance in cases with overlapping subproblems, although
    in the standard Combination Sum problem, the overlap isn't always significant.

    Time Complexity: O(N^(T/M)), where N is the number of candidates, T is the target,
                    and M is the minimum value in candidates.  In the worst case, it's the same
                    as the basic recursive approach, but with memoization, it can be faster
                    for certain inputs due to avoiding redundant calculations.
    Space Complexity: O(T * (T/M)) -  The size of the memoization dictionary.
    """
    memo = {}  # Store results of subproblems: (remaining, start) -> list of combinations

    def backtrack(remaining, start):
        if (remaining, start) in memo:
            return memo[(remaining, start)]

        if remaining == 0:
            return [[]]  # Base case: one empty combination
        if remaining < 0:
            return []

        result = []
        for i in range(start, len(candidates)):
            sub_combinations = backtrack(remaining - candidates[i], i)
            for comb in sub_combinations:
                result.append([candidates[i]] + comb)  # Prepend current candidate
        memo[(remaining, start)] = result  # Store the result
        return result

    return backtrack(target, 0)



if __name__ == "__main__":
    candidates = [2, 3, 6, 7]
    target = 7

    print(f"Candidates: {candidates}, Target: {target}")
    print("\n1. Recursive Approach:")
    result1 = combinationSum_recursive(candidates, target)
    print(result1)

    print("\n2. Optimized Recursive Approach with Sorting:")
    result2 = combinationSum_optimized(candidates, target)
    print(result2)

    print("\n3. Iterative Approach using a Stack:")
    result3 = combinationSum_iterative(candidates, target)
    print(result3)

    print("\n4. Dynamic Programming Approach:")
    result4 = combinationSum_dp(candidates, target)
    print(result4)

    print("\n5. Generator-Based Approach:")
    result5 = combinationSum_generator(candidates, target)
    print(result5)

    print("\n6. Memoized Recursive Approach:")
    result6 = combinationSum_memoized(candidates, target)
    print(result6)
